src/lib/waku_message/version_1.ts

```diff
@@ -1,100 +1,271 @@
 const PaddingTarget = 256;
 const SignatureLength = 65;
+const OneMillion = BigInt(1_000_000);
+
+export const Version = 1;
 
 export type Signature = {
   signature: Uint8Array;
   publicKey: Uint8Array | undefined;
 };
 
-/**
- * Encode the payload pre-encryption.
- *
- * @internal
- * @param messagePayload: The payload to include in the message
- * @param sigPrivKey: If set, a signature using this private key is added.
- * @returns The encoded payload, ready for encryption using {@link encryptAsymmetric}
- * or {@link encryptSymmetric}.
- */
-export async function clearEncode(
-  messagePayload: Uint8Array,
-  sigPrivKey?: Uint8Array
-): Promise<{ payload: Uint8Array; sig?: Signature }> {
-  let envelope = new Uint8Array([0]); // No flags
-  envelope = addPayloadSizeField(envelope, messagePayload);
-  envelope = concat([envelope, messagePayload]);
-
-  // Calculate padding:
-  let rawSize =
-    FlagsLength +
-    computeSizeOfPayloadSizeField(messagePayload) +
-    messagePayload.length;
-
-  if (sigPrivKey) {
-    rawSize += SignatureLength;
-  }
+export enum Method {
+  Asymmetric = "asymmetric",
+  Symmetric = "symmetric",
+}
 
-  const remainder = rawSize % PaddingTarget;
-  const paddingSize = PaddingTarget - remainder;
-  const pad = randomBytes(paddingSize);
+export interface CipherParams {
+  method: Method;
+  /**
+   * - private key if decrypting and { @link method } is { @link Method.Asymmetric }
+   * - public key if encrypting and { @link method } is { @link Method.Asymmetric }
+   * - symmetric key if encrypting or decrypting and { @link method } is { @link Method.Symmetric }
+   */
+  key: Uint8Array;
+}
 
-  if (!validateDataIntegrity(pad, paddingSize)) {
-    throw new Error("failed to generate random padding of size " + paddingSize);
+export class MessageV1 extends MessageV0 implements Message {
+  private readonly _decodedPayload: Uint8Array;
+
+  constructor(
+    proto: proto.WakuMessage,
+    decodedPayload: Uint8Array,
+    public signature?: Uint8Array,
+    public signaturePublicKey?: Uint8Array
+  ) {
+    super(proto);
+    this._decodedPayload = decodedPayload;
   }
 
-  envelope = concat([envelope, pad]);
-  let sig;
-  if (sigPrivKey) {
-    envelope[0] |= IsSignedMask;
-    const hash = keccak256(envelope);
-    const bytesSignature = await sign(hash, sigPrivKey);
-    envelope = concat([envelope, bytesSignature]);
-    sig = {
-      signature: bytesSignature,
-      publicKey: secp.getPublicKey(sigPrivKey, false),
-    };
+  get decodedPayload(): Uint8Array {
+    return this._decodedPayload;
   }
-
-  return { payload: envelope, sig };
 }
 
-/**
- * Decode a decrypted payload.
- *
- * @internal
- */
-export function clearDecode(
-  message: Uint8Array
-): { payload: Uint8Array; sig?: Signature } | undefined {
-  const sizeOfPayloadSizeField = getSizeOfPayloadSizeField(message);
-  if (sizeOfPayloadSizeField === 0) return;
+export class EncoderV1 implements Encoder {
+  constructor(
```

Comment:
`EncoderRLN` could take an encoder as input and override the `encoderProto` method to decode the protobuf `proof` field and provide added method to verify the proof.